x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/TransportPutSamlServiceProviderAction.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.idp.action;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.DocWriteResponse;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.ValidationException;
+import org.elasticsearch.common.hash.MessageDigests;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.util.iterable.Iterables;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.idp.saml.sp.SamlServiceProviderDocument;
+import org.elasticsearch.xpack.idp.saml.sp.SamlServiceProviderIndex;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Clock;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.stream.Collectors;
+
+public class TransportPutSamlServiceProviderAction
+    extends HandledTransportAction<PutSamlServiceProviderRequest, PutSamlServiceProviderResponse> {
+
+    private final Logger logger = LogManager.getLogger();
+    private final SamlServiceProviderIndex index;
+    private final Clock clock;
+
+    @Inject
+    public TransportPutSamlServiceProviderAction(TransportService transportService, ActionFilters actionFilters,
+                                                 SamlServiceProviderIndex index) {
+        this(transportService, actionFilters, index, Clock.systemUTC());
+    }
+
+    TransportPutSamlServiceProviderAction(TransportService transportService, ActionFilters actionFilters,
+                                                 SamlServiceProviderIndex index, Clock clock) {
+        super(PutSamlServiceProviderAction.NAME, transportService, actionFilters, PutSamlServiceProviderRequest::new);
+        this.index = index;
+        this.clock = clock;
+    }
+
+    @Override
+    protected void doExecute(Task task, final PutSamlServiceProviderRequest request,
+                             final ActionListener<PutSamlServiceProviderResponse> listener) {
+        final SamlServiceProviderDocument document = request.getDocument();
+        if (document.docId != null) {
+            listener.onFailure(new IllegalArgumentException("request document must not have an id [" + document.docId + "]"));
+            return;
+        }
+        index.findByEntityId(document.entityId, ActionListener.wrap(matchingDocuments -> {
```

Comment:
we could export document.entityId in a local var here and call `deriveDocumentId` with that too below as this is the only field we use from the document

Reply:
Sorry, I don't follow. I think you're implying that if I set the entityId into a local that will allow me to achieve something else that's benefitial, but I'm not sure what that something else is.

Reply:
yeah that was badly written. I actually mixed up `document.entityId` and document.docId` here and my thinking was that since you : 
- use `document.entityId` a few times already and
- you pass `document` in `deriveDocumentId()` but only use the `docId` in there

Do a 

`final String someName = document.entityId ;`

and use `someName` instead of `document.entityId` in all cases below and to pass to `deriveDocumentId`


It was a nit, I thought it reads better (in a similar manner to https://github.com/elastic/elasticsearch/pull/52886#pullrequestreview-366424333) , feel free to disregard if you don't agree that it does - I really do not hold strong opinions over these.

Reply:
I personally perfer _not_ assigning it to a local because I find it easier to read the code if it's clear that a particular value always comes from the object.

That is, I see `document.entityId` and, so long as I know what `document` is, I know that we're talking about the entityId of the document.  
However, if I assigned it to `docEntityId` I'm left wondering _why_ it's a local and if it's assigned from multiple places, etc, and need to navigate to where it's assigned to work out what it really is.

But I'm very interested in why you find the reverse to be true. But perhaps that's a conversation for another time.